xarray/backends/zarr.py

```diff
@@ -312,40 +313,99 @@ def encode_variable(self, variable):
+        check_encoding_set : list-like
+            List of variables that should be checked for invalid encoding
+            values
+        writer : ArrayWriter
+        unlimited_dims : list-like
+            List of dimension names that should be treated as unlimited
+            dimensions.
+            dimension on which the zarray will be appended
+            only needed in append mode
+        """
+
+        variables, attributes = self.encode(variables, attributes)
+
+        self.set_attributes(attributes)
+        self.set_dimensions(variables, unlimited_dims=unlimited_dims)
+        self.set_variables(variables, check_encoding_set, writer,
+                           unlimited_dims=unlimited_dims)
 
     def sync(self):
         pass
 
+    def set_variables(self, variables, check_encoding_set, writer,
+                      unlimited_dims=None, append_dim=None):
+        """
+        This provides a centralized method to set the variables on the data
+        store.
+
+        Parameters
+        ----------
+        variables : dict-like
+            Dictionary of key/value (variable name / xr.Variable) pairs
+        check_encoding_set : list-like
+            List of variables that should be checked for invalid encoding
+            values
+        writer :
+        unlimited_dims : list-like
+            List of dimension names that should be treated as unlimited
+            dimensions.
+        append_dim: str
+            dimension on which the zarray will be appended
+            only needed in append mode
+        """
+        for vn, v in variables.items():
+            name = _encode_variable_name(vn)
+            check = vn in check_encoding_set
+            attrs = v.attrs.copy()
+            dims = v.dims
+            dtype = v.dtype
+            shape = v.shape
+
+            fill_value = attrs.pop('_FillValue', None)
+            if v.encoding == {'_FillValue': None} and fill_value is None:
+                v.encoding = {}
+            append = False
+            try:
+                zarr_array = self.ds[name]
+                append = True
+            except KeyError:
+                encoding = _extract_zarr_variable_encoding(
+                    v, raise_on_invalid=check)
+                encoded_attrs = OrderedDict()
+                # the magic for storing the hidden dimension data
+                encoded_attrs[_DIMENSION_KEY] = dims
+                for k2, v2 in attrs.items():
+                    encoded_attrs[k2] = self.encode_attribute(v2)
+
+                zarr_array = self.ds.create(name, shape=shape, dtype=dtype,
+                                            fill_value=fill_value, **encoding)
+                zarr_array.attrs.put(encoded_attrs)
+                zarr_array[...] = v.data
+            if append:
+                if self.append_dim is None:
+                    raise ValueError('The dimension on which the data is \
+                     appended has to be named.')
```

Comment:
What if we just want to add a new variable to an existing zarr store. This PR could hypothetically support that case as well, but in that case, there is no append_dim to specify.

Reply:
As mentioned in the other comment, it should already work, but I will add another test for it :+1: